ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PeerSimulator/Tests/Rollback.hs

```diff
@@ -0,0 +1,85 @@
+{-# LANGUAGE BlockArguments      #-}
+{-# LANGUAGE DerivingStrategies  #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Test.Consensus.PeerSimulator.Tests.Rollback (tests) where
+
+import           Control.Monad.IOSim (runSimOrThrow)
+import           Data.Maybe (fromJust)
+import           Ouroboros.Consensus.Block (ChainHash (..))
+import           Ouroboros.Consensus.Config.SecurityParam
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import           Test.Consensus.BlockTree (BlockTree (..), BlockTreeBranch (..))
+import           Test.Consensus.Genesis.Setup
+import           Test.Consensus.PeerSimulator.Run (noTimeoutsSchedulerConfig)
+import           Test.Consensus.PeerSimulator.StateView
+import           Test.Consensus.PointSchedule
+import qualified Test.QuickCheck as QC
+import           Test.QuickCheck
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.TestBlock (TestBlock, unTestHash)
+
+tests :: TestTree
+tests = testGroup "rollback" [
+  testProperty "can rollback" (prop_rollback True),
+  testProperty "cannot rollback" (prop_rollback False)
+  ]
+
+-- | The boolean parameter specifies whether we are interested in the property
+-- when we can rollback.
+prop_rollback :: Bool -> QC.Gen QC.Property
```

Comment:
I think the documentation should explain with more precision what is the property being tested. I don't get it yet.

Reply:
The goal for this is to be a sanity check for the PeerSimulator, not a proper consensus test (hence the location in `Consensus.PeerSimulator.Tests.*` and not `Consensus.Genesis.Tests.*` or something). It just aims at checking that the rollback is properly implemented by the mocked ChainSync server and properly followed by the client. Shall I add this to this property? Or maybe the whole module?

Reply:
I'd like it to say something like
```
@prop_rollback True@ tests that the selection of the node under test changes branches when sent a rollback to a block no older than k blocks before the current selection.

@prop_rollback False@ tests that ...
```

Reply:
Sounds great! Will do that.

Reply:
WDYT about 291df5f1d?

Reply:
That helps. Thanks!